### PR TITLE
Fix for #314

### PR DESCRIPTION
**Pull request by yyny**
Since `w->opacity`, `w->opacity_target` and `win_calc_opacity_target(ps, w, true)` are always less than `1`.

### CODE REVIEW

**Comment by jransijn** [comment]
I didn't know we had this tool, pretty cool. Once again though, would be nice if we had regression tests to catch bugs like this so they never happen again. The [new config system](https://github.com/jransijn/picom/blob/modules/src/utils/cfg.h) I'm working on could maybe at some point be used to check many different configurations at once.

File: src/backend/backend.c [hunk 1]
```diff
@@ -204,15 +204,17 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
 			if (w->state == WSTATE_MAPPING) {
 				// Gradually increase the blur intensity during
 				// fading in.
-				blur_opacity = w->opacity / w->opacity_target;
```
Comment (yshui) [inline src/backend/backend.c]: The intention here is that as `w->opacity` gradually goes from `0` to `w->opacity_target`, `blur_opacity` gradually increase to 1. I don't understand what using multiply here means.

File: src/backend/backend.c [hunk 2]
```diff
@@ -204,15 +204,17 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
...
+				blur_opacity = w->opacity * w->opacity_target;
 			} else if (w->state == WSTATE_UNMAPPING ||
 			           w->state == WSTATE_DESTROYING) {
 				// Gradually decrease the blur intensity during
 				// fading out.
 				blur_opacity =
-				    w->opacity / win_calc_opacity_target(ps, w, true);
```
Comment (yshui) [inline src/backend/backend.c]: Same as above. Here `w->opacity` goes from `win_calc_opacity_target(ps, w, true)` to `0`.

File: src/backend/backend.c (hunk 1 above)
Comment (yshui) [reply src/backend/backend.c]: `w->opacity` should not exceed `w->opacity_target`, if it does, there is a bug somewhere else in my code.

Comment (jransijn) [reply src/backend/backend.c]: Ah, I didn't read the comments correctly, I thought that code was meant to *decrease* the blur opacity. I think the real problem is that `opacity_target` exceeds `opacity`, then, yes.

Comment (jransijn) [reply src/backend/backend.c]: `Assertion `w->opacity <= w->opacity_target' failed.`

So yes, that is the problem. It takes some effort to trigger it, though... Maybe a floating point error?

Comment (jransijn) [reply src/backend/backend.c]: So perhaps a hint:

```
[ 02/23/2020 22:45:41.555 paint_all_new TRACE ] 0.870000 <= 1.000000
BLAH BLAH BLAH
[ 02/23/2020 22:45:41.573 paint_all_new TRACE ] 0.810000 <= 0.625000

picom: ../src/backend/backend.c:208: paint_all_new: Assertion `w->opacity <= w->opacity_target' failed.
```

It looks like `opacity_target` is changed while `opacity` is still lower than the new value.

Comment (jransijn) [reply src/backend/backend.c]: Ah, I think I understand the problem...

```
$ cat ~/.xmonad/xmonad.hs | grep 0.375
        [ isUnfocused --> transparency 0.375
```

`0.625` is the default opacity of unfocused windows on my setup. It looks like picom incorrectly assumes the opacity of an unmapped/mapping window can never change.

As for a quick solution:

I think it is reasonable to just jump to the `opacity_target` when we detect an external program changed the opacity during a mapping. This should be rare, and is probably the 'intended' opacity in that case.

Comment (jransijn) [reply src/backend/backend.c]: This doesn't solve the problem for unmapping windows, though. Also, I think the problem is really somewhere else in the code

Comment (jransijn) [reply src/backend/backend.c]: Also, while I was testing, I got this:
```
Assertion `(w->flags & WIN_FLAGS_SHADOW_STALE) == 0' failed.
```

Yeah, I'm not getting into that right now. I'll save the log, though

Comment (tryone144) [reply src/backend/backend.c:207]: > It looks like picom incorrectly assumes the opacity of an unmapped/mapping window can never change.

This is a problem as it certainly can change by changing focus (`inactive-opacity`) or if set directly, e.g. via `picom-trans`.

> As for a quick solution:
> 
> I think it is reasonable to just jump to the `opacity_target` when we detect an external program changed the opacity during a mapping. This should be rare, and is probably the 'intended' opacity in that case.

`blur_opacity` is intended to be 1, if the window is `MAPPED` (and visible) and not directly dependent on window opacity EXCEPT when fading in/out. So clamping to 1 should be more reasonable.

**Pull request closed without merging**